src/rules/export.js

```diff
@@ -24,6 +24,24 @@ ambient namespaces:
 const rootProgram = 'root'
 const tsTypePrefix = 'type:'
 
+/**
+ * Detect function overloads like:
+ * ```ts
+ * export function foo(a: number);
+ * export function foo(a: string);
+ * export function foo(a: number|string) { return a; }
+ * ```
+ * @param {Set<Object>} nodes
+ * @returns {boolean}
+ */
+function isTypescriptFunctionOverloads(nodes) {
+  return Array.from(nodes).some(node => node.parent.type === 'TSDeclareFunction') &&
+    Array.from(nodes).every(node => (
```

Comment:
let's definitely not do Array.from twice. maybe something like this for the body of the function?
```js
const types = new Set(Array.from(nodes, node => node.parent.type));
if (!types.has('TSDeclareFunction') || types.size > 2) {
  return false;
}
return types.size === 1 || types.has('FunctionDeclaration');
```

Reply:
I end up with this:
```js
const types = new Set(Array.from(nodes, node => node.parent.type))
return types.size === 2 && types.has('TSDeclareFunction') && types.has('FunctionDeclaration')
```
The case when `types` is the same as `new Set("TSDeclareFunction")` is when the function lacks implementation, which is a tsc error.

Reply:
That isn't the same logic though - your current logic allows it to be only "TSDeclareFunction". Functions don't need an implementation in tsc if they're just defining a type overload.

Reply:
That case is a tsc error https://www.typescriptlang.org/play/#code/KYDwDg9gTgLgBAMwK4DsDGMCWEWIBQCUQA . And for the purpose of this rule, that case does not matter.